### ISSUE TITLE
consider running go generate

### ISSUE THREAD

**Issue by marten-seemann** [protocol/.github#123]
Should we run `go generate` in CI (probably in `go-check`)?
In general, there's no reason why generated code should not be up to date, _assuming_ that `go generate` is deterministic.

Not sure if that's a valid assumption. For quic-go it isn't, as we [generate](https://github.com/lucas-clemente/quic-go/blob/master/fuzzing/header/cmd/corpus.go) the fuzzing corpus there with a random input, but maybe that's a bug in quic-go that should be fixed?

I'm raising this issue because I just created a [PR](https://github.com/multiformats/go-multicodec/pull/49) in the multicodec repo that adds a separate workflow for that.

**Comment by mvdan** [comment]
I don't think you can generally assume that `go generate` is deterministic. Here are some examples:

1) Lines like `//go:generate stringer args...`, which don't specify the version of the tool in question. https://github.com/golang/go/issues/42088 will fix this, but 1.17 is still out by a couple of months, and it will be another few months until most software at PL pins versions of generate tools. There's always `//go:generate go run golang.org/x/tools/cmd/stringer args...`, but that requires requiring a version in go.mod, which you might not want to do.

2) Like point 1, but when tools aren't fully in Go, such as with protoc. You often have to give up and just document that people should use a specific version, often enforced via e.g. `make`.

3) Generators which are reasonably deterministic within a specific point in time only, such as those that fetch input from an HTTPS URL. You can fairly confidently verify that a PR authored today has up-to-date generated code, but if master sits there for a few weeks, it will likely need re-generation by design. One could argue that a specific version of the data should be pinned via the URL, but sometimes that's not possible.

And then there's generators which are deliberately non-deterministic or random - I do think those shouldn't use `go generate`, because then it's almost always impossible to use `go generate ./...` without inserting more noise into the git log.

I think point 1 will improve soon with Go 1.17, but points 2 and 3 will continue to be a nuisance. For point 2, CI will likely not have the right version of the tools in question, or not have the tool installed at all. For point 3, you're going to risk being noisy and encouraging devs to stop using the checks. You can certainly do your best to keep your `go generate` lines deterministic, and I generally do, but I don't think we can count on that for all of PL.

All in all, I can't think of a general way to do this. It will certainly work and be good for some repos though. Maybe we make it opt-in as part of the config, like other future "extra sanity checks" that might not make sense for some repos.
